File: ConvertSpool2h5.py
#!/usr/bin/env python
"""
Michael Hirsch
Sept 2015

Streams data from  spool file input disk to HDF5 output disk (both USB HDD) at 35 MB/s. Did not attempt to optimize.

Example performance: 4128 spool files selected by Detect.py, each with 47 320x270 images  33.5 GB -> HDF5 21.6 GB compression=1 in 948.5 sec.
That's 35 MB/sec.

# Example for CV_ionosphere automatic auroral detection output
1. ./FileTick.py  z:\2017-04-04a\spool
2. ../cv_ionosphere/Detect.py z:\2017-04-0a\spool\index.h5 z:\2017-04-27\ dmc2017.ini -k 10
2a. (optional patch for bad detect) ../cv_ionosphere/PatchAuroraldet.py
3. ./ConvertSpool2h5.py z:\2017-04-04a\spool\index.h5 -det z:\2017-04-04a\auroraldet.h5

./ConvertSpool2h5.py ~/H/neo2012-12-25/spool_5/index.h5 -det ~/data/2012-12-25/auroraldet.h5 -o ~/data/2012-12-25/extracted.h5 -xy 320 270 -stride 648 -z 4
or to convert all spool files without regard to detections
./ConvertSpool2h5.py ~/H/neo2012-12-25/spool_5/index.h5 -o ~/data/2012-12-25/extracted.h5 -xy 320 270 -stride 648 -z 4
"""
import logging
from time import time
from sys import argv
from dateutil.parser import parse
from pathlib import Path
import h5py
from pandas import read_hdf
import numpy as np
#
from histutils import vid2h5
from dmcutils.neospool import oldspool,h5toh5, readNeoSpool,spoolparam

W = 51  # keep +/-  W/2 frames around detection

def converter(p):
    tic = time()

    P = {'kineticsec':p.kineticsec,'rotccw':p.rotccw,'transpose':p.transpose,
         'flipud':p.flipud,'fliplr':p.fliplr,'fire':p.fire,
             }

    path = Path(p.path).expanduser()

    if path.is_file() and path.suffix == '.h5' :
        with h5py.File(path, 'r') as f:
            tickfile = 'filetick' in f  # spool file index

        if tickfile:
            """
            1. read index file
            2. (optional) read detection file
            3. convert specified file to HDF5
            """
            spoolini = path.parent / 'acquisitionmetadata.ini'
            if p.startutc is None:
                tstart = spoolini.stat().st_ctime  # crude measure of start time
            else:
                tstart = parse(p.startutc).timestamp()
            assert isinstance(tstart,(float,int))
# %% 1.
            flist = read_hdf(path,'filetick')
# %% 2. select which file to convert...automatically
            if p.detfn:
                detfn = Path(p.detfn).expanduser()
                with h5py.File(detfn,'r',libver='latest') as f:
                    det = f['/detect'][:]

                upfact = flist.shape[0] // det.size

                if not 1 <= upfact <= 20:
                    logging.error(f'was {detfn} sampled correctly?')
                    return

                det2 = np.zeros(flist.shape[0])
                try:
                    det2[:det.size*upfact-1:upfact] = det  # gaps are zeros
                except ValueError: # off by one
                    det2[:det.size*upfact:upfact] = det  # gaps are zeros

                assert abs(len(flist) - det2.size) <= 20,f'{detfn} and {path} are maybe not for the same spool data file directory'
                det = det2

                Lkeep = np.ones(min(len(flist),W),dtype=int)  # keeps Lkeep/2 files each side of first/last detection.

                ikeep = np.convolve(det,Lkeep,'same').astype(bool)
                det = det[ikeep]

                assert len(det) == ikeep.sum(),f'len(det): {len(det)}  ikeep.sum(): {ikeep.sum()}'
                assert len(ikeep)  == len(flist), f'len(flist): {len(flist)} len(ikeep): {len(ikeep)}'
            else:  # just convert all spool files (can be terabyte+ of data into HDF5)
                print('no detection file specified, converting all Spool files')
                det = np.ones(flist.shape[0],dtype=bool)
                ikeep = slice(None)
# %% 3.
            Fparam = spoolparam(spoolini,
                                p.xy[0]//p.bin[0], p.xy[1]//p.bin[1], p.stride)
            P = {**P,**Fparam}

            flist2 = flist[ikeep]
            print(f'keeping/converting {flist2.shape[0]} out of {flist.shape[0]} files in {path}')

            # append to HDF5 one spool file at a time to conserve RAM
            for i,fn in enumerate(flist2):
                fn = Path(path.parent/fn)
                P['spoolfn'] = fn
                imgs, ticks, tsec = readNeoSpool(fn, P, zerocols=p.zerocols)
                vid2h5(imgs, None, None, ticks, p.outfn, P, argv, i, len(flist2), det, tstart)
        else:
            print('writing metadata')
            rawind,ut1_unix = h5toh5(path, p.kineticsec, p.startutc)
            vid2h5(None, ut1_unix, rawind, p.outfn, P)
    elif p.broken:  # ancient spool file < 2011
        rawind,ut1_unix = oldspool(path, p.xy, p.bin, p.kineticsec, p.startutc, p.outfn)
        vid2h5(None, ut1_unix, rawind, p.outfn, P)
    else:
        raise ValueError('Not sure what to do with your specified options')

    print(f'wrote {p.outfn} in {time()-tic:.1f} sec.')

if __name__ == "__main__":
    from argparse import ArgumentParser
    p = ArgumentParser(description='Andor Neo Spool reader, plotter, converter')
    p.add_argument('path',help='path containing 12-bit Neo spool files in broken format (2008-spring 2011)')
    p.add_argument('-detfn',help='path to detections.h5 file')
    p.add_argument('-xy',help='nx ny  number of x and y pixels respectively',nargs=2,default=(2544,2160),type=int)
    p.add_argument('-b','--bin',help='nx ny  number of x and y binning respectively',nargs=2,default=(1,1),type=int)
    p.add_argument('-k','--kineticsec',help='kinetic rate of camera (sec)  = 1/fps',type=float)
    p.add_argument('--rotccw',help='rotate CCW value in 90 deg. steps',type=int,default=0)
    p.add_argument('--transpose',help='transpose image',action='store_true')
    p.add_argument('--flipud',help='vertical flip',action='store_true')
    p.add_argument('--fliplr',help='horizontal flip',action='store_true')
    p.add_argument('-s','--startutc',help='utc time of nights recording')
    p.add_argument('-o','--outfn',help='extract raw data into this file [h5,fits,mat]')
    p.add_argument('-v','--verbose',help='debugging',action='count',default=0)
    p.add_argument('-stride',help='length of footer in spool file',type=int)
    p.add_argument('-z','--zerocols',help='number of zero columns in spool file',type=int,default=8)
    p.add_argument('-fire',help='fire filename')
    p.add_argument('-broken',help='enables special Matlab reader for broken 2009-2010 spool files',action='store_true')
    p = p.parse_args()

    converter(p)

File: ConvertSpool2h5.m
function ConvertSpool2h5(varargin)
% given a Neo 12bit packed spool directory (with the corrupted 2008 Andor Solis
% beta files) read all the images in that directory and convert to FITS
% That is, very many .dat spool files from a single directory are converted to
% one big FITS file, readable in ImageJ, etc.
% Michael Hirsch
%
% CAVEAT: h5create() does not allow for compound dataset create, you have
% to use low-level commands

p = inputParser;
addRequired(p,'spooldir')
addOptional(p,'outfn',[])
addRequired(p,'startutc')
addRequired(p,'kineticsec')
addOptional(p,'nx',2544)
addOptional(p,'ny',2160)
addOptional(p,'lla',[65.1186367, -147.432975, 500.])
parse(p,varargin{:})
U = p.Results;

datlist = dir([U.spooldir,'/*.dat']);
datfn = sort({datlist.name});
disp(['I found ',int2str(length(datfn)),' .dat files in ',U.spooldir])
%% auto output filename
if U.spooldir(end) == pathsep
    full = U.spooldir(1:end-1);
else
    full = U.spooldir;
end
[~,basename] = fileparts(full);
if isempty(U.outfn)
    outfn = expanduser([U.spooldir,'/',basename,'.h5']);
else
    outfn = expanduser(U.outfn);
end
disp(['writing to ',outfn])
%% read
nfiles = length(datfn);
%data = zeros([ny,nx,nfiles],'uint16');

h5create(outfn,'/rawimg',[U.nx,U.ny,nfiles],'Datatype','uint16',...
         'Deflate',6,'Chunksize',[U.nx,U.ny,1],'shuffle',true)
for i = 1:nfiles
    data = readNeoPacked12bit([U.spooldir,'/',datfn{i}],U.nx,U.ny);
    disp([int2str(i),' / ',int2str(nfiles)])  
    
    h5write(outfn,'/rawimg',transpose(data),[1,1,i],[U.nx,U.ny,1])
end
%% output hdf5

h5writeatt(outfn,'/rawimg','CLASS',        'IMAGE')
h5writeatt(outfn,'/rawimg','IMAGE_VERSION','1.2')
h5writeatt(outfn,'/rawimg','IMAGE_SUBCLASS','IMAGE_GRAYSCALE')
h5writeatt(outfn,'/rawimg','DISPLAY_ORIGIN','LL')
h5writeatt(outfn,'/rawimg','IMAGE_WHITE_IS_ZERO',uint8(0))
%
h5create(outfn,'/rawind',nfiles,'Datatype','int64')
h5write(outfn,'/rawind',int64(1:nfiles))

t0 = datetime(U.startutc);
epoch = datetime(1970,1,1,0,0,0);
t= seconds(t0-epoch) + (0:(nfiles-1))*U.kineticsec;

h5create(outfn,'/ut1_unix',nfiles)
h5write(outfn,'/ut1_unix',t)
%
h5create(outfn,'/sensorloc/glat',1,'Datatype','double')
h5write(outfn,'/sensorloc/glat',U.lla(1))
h5create(outfn,'/sensorloc/glon',1,'Datatype','double')
h5write(outfn,'/sensorloc/glon',U.lla(2))
h5create(outfn,'/sensorloc/alt_m',1,'Datatype','double')
h5write(outfn,'/sensorloc/alt_m',U.lla(3))
%
h5create(outfn,'/params/kineticsec',1,'Datatype','double')
h5write(outfn,'/params/kineticsec',U.kineticsec)

end
